### ISSUE TITLE
download_features returns many incomplete .zip files

### ISSUE THREAD

**Issue by massimozanetti** [SDFIdk/CDSETool#67]
Hi. and thank you for develping CDSEtool, it is something really important.

I have tried it to download some Sentinel-2 products, and after a couple of runs I discovered that many products are not complete after download, and this happens randomly. See pictures two different instances of downloading the same list of products:

run 1
![brokenzip](https://github.com/SDFIdk/CDSETool/assets/129863715/c73e94d9-ee46-448e-866f-2557405d8535)

run 2
![brokenzip2](https://github.com/SDFIdk/CDSETool/assets/129863715/06757256-fc53-4b8e-83e7-859067e46ae0)

Many of the m are clearly incomplete, and they are not the same in the two runs...
Thanks

**Comment by pedep** [comment]
Hi @massimozanetti,

Thank you for your bug report
This is definitely a critical problem that should be adressed immediately :skull: 

I will try to reproduce your issue to see if something can be done. 
If i am lucky, there is some metadata in the headers from the download server, which will allow us to detect incomplete downloads. If not, i might have to quickly implement #40 in hopes of solving the issue.

In the meantime, could i get you to share a bit about the environment you are running `cdsetool` in?
- Operating system & version
- Python version, package manager and environment (conda, pip, etc)
- cdsetool version (`pip show cdsetool` or similar)
- Which internet connection is used? (corporate, domestic, university, LTE, etc)
  - Is the internet connection wired or wireless?
- Do you frequently experience network issues or package loss?
- If possible, could you share the script you are running?
  - If you cannot share your script, could you create a minimal script that recreates the issue?

Thank you for your assistance

**Comment by massimozanetti** [comment]
Hi @pedep, I am running on:

- Ubuntu 18.04.6 LTS
- conda 4.14.0
- pip install cdsetool==0.2.3
- Wired Eduroam internet (institution: https://www.fbk.eu/en/)

I do not experience data loss, I strongly believe the problem is connected to the procedure documented by CDSE. The chuncked download with no checks returns this kind of problems if naively implemented as suggested here: https://documentation.dataspace.copernicus.eu/APIs/OData.html

To reproduce the problem it just takes reproducing your documented snippets, with the list of products I am interested in as input, here it is for your convenience:
```
from cdsetool.credentials import Credentials
from cdsetool.query import query_features
from cdsetool.download import download_features

username = ..........
password = .........
products_dir = .........

#uuid_list = ['f2b18bec-4c8b-5c34-b253-d3e881dabbf8', 'a329664e-afa6-5866-9dc3-6830377906bf', '327905fc-6cf7-55b8-905b-630dc4c66615', '3ea6870a-1061-5db3-854a-a1ea1089ecf9', 'a8691ced-c685-5abb-86c9-f74d822e3b46', 'e3a6273e-d5cb-5f05-88e6-2afbbadb6cfe', '4b36b3c6-1d1c-5c51-91c5-5e3d77de6f3b', '807fd94e-8509-5d55-b1d4-6dfb96beced5', 'f233bb25-d79c-50bf-aca7-b3f7d27b7f83', '5628329c-9899-5814-ac41-aff14fa75443', '15810eaf-f5b5-50bf-89d0-32cf895ca18c', 'bf19b5c3-f0c4-5191-a638-a5ba94160693', '540a7b48-2d78-5ed5-996f-ad81a08ef628', '15a655d3-38c1-59d3-85ba-895afc2819cf', 'f62c9a83-63ba-5e02-9e0b-27cd3f128321', 'af723bbf-513f-5d25-8707-2e093a1a0ad5', 'e29669ab-2d25-5258-8ff6-cdb0c70e3e49', 'ba0f80ab-7acc-5777-8ebd-ce86edcfaad5', 'b86fe089-b65d-5c27-8b7e-c73951009c14', 'e88c4152-4a1f-57f6-bc02-e1cd33eb66de']

#query features
features_list = []
for uuid in uuid_list:
  qfeatures = query_features('Sentinel2',{'identifier': uuid})
  for feature in qfeatures:
    features_list.append(feature)

#istantiate credentials
credentials = Credentials(username, password)
#download products
downloads = download_features(features_list, products_dir, {'concurrency': 4, 'credentials': credentials})

for xid in downloads:
  print(f"feature {xid} downloaded")
```

**Comment by pjonsson** [comment]
I don't have a reproducer, but I also see faulty zip files which I assume is caused by this issue. I'm using a wired connection, which I (based on failures of other things) believe has intermittent issues.

Implementing #40 should not be necessary to detect the issue I'm seeing, trying to unzip the file will reveal that it is broken.

@pedep what is https://github.com/SDFIdk/CDSETool/blob/8385e3a82d0915f472458955608ec40aadb0e082/src/cdsetool/download.py#L52 intended to avoid?

And I think https://github.com/SDFIdk/CDSETool/blob/8385e3a82d0915f472458955608ec40aadb0e082/src/cdsetool/download.py#L50 looks suspicious, it opens a file that we are already holding an open file descriptor to. Why isn't the file descriptor passed to open? I have no idea if `mkstemp` opens the file with `closefd=True`, but unless it does, the line `os.close(fd)` could be removed if the file descriptor was passed to open.

**Comment by pedep** [comment]
@pjonsson I read a few articles regarding streaming to files in python, and a lot of them used a similar syntax

I remember someone explaining that reqeusts might return `None` or something like that in some cases due to the network misbehaving, which is why we should filter them out.
I thought that seemed reasonable enough, and decided not to go diving into the sourcecode for `requests` for answers.

@massimozanetti #90 attempts to solve this problem, and will be released in version 0.2.5 soon.

**Comment by pjonsson** [comment]
I don't think my PR solves this issue, but it appears to improve the situation.

**Comment by pjonsson** [comment]
@massimozanetti are you using urllib3 version 1.x or 2.x?

**Comment by massimozanetti** [comment]
@pjonsson I am running on Conda with urllib3 v1.26.7

**Comment by pjonsson** [comment]
@massimozanetti you provided a reproducer in https://github.com/SDFIdk/CDSETool/issues/67#issuecomment-1908043776, could you run that on your network with CDSETool 0.2.6 and see if the issue is still reproducible?

When I upgraded to urllib3 2.x, my issues changed. I cannot tell if that is caused by the upgrade or some external factor (network health, Copernicus server status, etc), but if it is easy for you, you could try to upgrade to urllib3 and see if that makes any difference.

**Comment by massimozanetti** [comment]
@pjonsson I've run the reproducer again, as you asked, with: 
```
cdsetool==0.2.6
urllib3==1.26.18
```
Now, still some of the products are incomplete, but differently from the previous attempt, running it a few times the incomplete ones are always the same! So we have now lost the randomness element to the issue. But an issue is still there. 

If you also run the same snippet I've previously shared, can you check if you obtain the same details as here in the figure?
![1](https://github.com/SDFIdk/CDSETool/assets/129863715/1d44d498-b009-451c-9e93-a53c4bfd2349)

**Comment by pjonsson** [comment]
@massimozanetti not having randomness is a great step forward! Do you always get incomplete products, or only sometimes? Can you try a sequential version, so replacing:
```python
#istantiate credentials
credentials = Credentials(username, password)
#download products
downloads = download_features(features_list, products_dir, {'concurrency': 4, 'credentials': credentials})

for xid in downloads:
  print(f"feature {xid} downloaded")
```
with:
```python
options = {'credentials': Credentials(username, password)}
#download products
for feature in features_list:
  xid = download_feature(features_list, products_dir, options)
  print(f"feature {xid} downloaded")
```
and see if you still get incomplete products?

Since I get partial zip files as well, I want to help us both, but for various reasons relating to my setup, I can't run your reproducer easily.

**Comment by massimozanetti** [comment]
Hi @pjonsson,

- sequential download (your snippet) gives me exactly the same incomplete data as in the bulk download runs;
- printout is ok for the case of bulk download (StatusMonitor() on), it lists file names and sizes;
- printout also ok for the sequential download (your snippet), it lists file names;

"None" is never seen

**Comment by pjonsson** [comment]
@massimozanetti Well, for better or worse, it's not the parallel downloads that are causing your issue. Can you try the sequential download but with urllib3 2.x, or is that difficult? (I'm hoping that will start throwing exceptions when you get the partial downloads.)

**Comment by pjonsson** [comment]
@pedep do you know enough Conda to help massimozanetti test this with urllib3 2.x?

**Comment by pedep** [comment]
Sure, heres a snippet from my terminal where i have the same version as @massimozanetti and am upgrading to latest
```
$ conda list | grep urllib
urllib3                   1.26.18         py312h06a4308_0

$ conda search urllib | grep -P "2\.\d\.\d"
urllib3                        2.0.3 py310h06a4308_0  pkgs/main
urllib3                        2.0.3 py311h06a4308_0  pkgs/main
urllib3                        2.0.3 py312h06a4308_0  pkgs/main
urllib3                        2.0.3  py38h06a4308_0  pkgs/main
urllib3                        2.0.3  py39h06a4308_0  pkgs/main
urllib3                        2.0.7 py310h06a4308_0  pkgs/main
urllib3                        2.0.7 py311h06a4308_0  pkgs/main
urllib3                        2.0.7  py38h06a4308_0  pkgs/main
urllib3                        2.0.7  py39h06a4308_0  pkgs/main
urllib3                        2.1.0 py310h06a4308_0  pkgs/main
urllib3                        2.1.0 py310h06a4308_1  pkgs/main
urllib3                        2.1.0 py311h06a4308_0  pkgs/main
urllib3                        2.1.0 py311h06a4308_1  pkgs/main
urllib3                        2.1.0 py312h06a4308_0  pkgs/main
urllib3                        2.1.0 py312h06a4308_1  pkgs/main
urllib3                        2.1.0  py38h06a4308_0  pkgs/main
urllib3                        2.1.0  py38h06a4308_1  pkgs/main
urllib3                        2.1.0  py39h06a4308_0  pkgs/main
urllib3                        2.1.0  py39h06a4308_1  pkgs/main

$ conda install urllib3==2.1.0
Channels:
 - defaults
Platform: linux-64
Collecting package metadata (repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /home/hexabit/miniconda3

  added / updated specs:
    - urllib3==2.1.0


The following packages will be UPDATED:

  urllib3                           1.26.18-py312h06a4308_0 --> 2.1.0-py312h06a4308_1


Proceed ([y]/n)? y


Downloading and Extracting Packages:

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
```

### TLDR
 `conda install urllib3==2.1.0`

**Comment by pedep** [comment]
As far as i can see, `> 2.1` is not present in the defaults channel, so either go with the conda-forge version `conda install conda-forge::urllib3==2.2.1` or go with the latest `< 2.1` version from defaults `conda install urllib3==2.0.7`

**Comment by pjonsson** [comment]
If it is Python 3.8 (or 3.9?) there's something about SSL that changed. But 2.0.7 is what I used until midnight, so use that. If there are ChunkedEncodingError exceptions, it's likely the same issue that I'm seeing.

**Comment by massimozanetti** [comment]
@pjonsson I have a bit of a problem before updating `urllib3==2.0.7`, the installer says:
```
The environment is inconsistent, please check the package plan carefully
The following packages are causing the inconsistency:

  - defaults/linux-64::geopandas==0.12.2=py38h06a4308_0                                                                                                          done
```
and
```
The following NEW packages will be INSTALLED:

  geopandas-base     pkgs/main/linux-64::geopandas-base-0.12.2-py38h06a4308_0
  pyproj             pkgs/main/linux-64::pyproj-2.6.1.post1-py38hb3025e9_1

The following packages will be UPDATED:

  certifi                         2023.11.17-py38h06a4308_0 --> 2024.2.2-py38h06a4308_0
  requests                            2.31.0-py38h06a4308_0 --> 2.31.0-py38h06a4308_1
  urllib3                            1.26.18-py38h06a4308_0 --> 2.0.7-py38h06a4308_0
```
The problem is that it is trying to DOWNGRADE my `geopandas` and `pyproj`, and I really need to have them updated to the latest version to not screw up other scripts I need to mantain...

**Comment by pedep** [comment]
I just installed latest geopandas, and it seems to work just fine with `urllib==2.0.7` on python 3.12

```
$ python3 --version
Python 3.12.1
$ conda list | grep "geopandas\|cdsetool\|urllib"
cdsetool                  0.2.7                    pypi_0    pypi
geopandas                 0.14.2          py312h06a4308_0
geopandas-base            0.14.2          py312h06a4308_0
urllib3                   2.0.7              pyhd8ed1ab_0    conda-forge
```

and i have a environment i downgraded to 3.8

```
$ conda list | grep "cdsetool\|geopandas\|urllib"
cdsetool                  0.2.7                    pypi_0    pypi
geopandas                 0.13.2                   pypi_0    pypi
geopandas-base            0.12.2           py38h06a4308_0
urllib3                   2.0.7            py38h06a4308_0

$ python --version
Python 3.8.18

**Comment by pedep** [comment]
Yes, that seems to work just fine too
```
$ python --version
Python 3.8.18

$ conda list | grep "cdsetool\|geopandas\|urllib\|pyproj"
cdsetool                  0.2.7                    pypi_0    pypi
geopandas                 0.13.2                   pypi_0    pypi
geopandas-base            0.12.2           py38h06a4308_0
pyproj                    3.5.0                    pypi_0    pypi
urllib3                   2.0.7            py38h06a4308_0

$ python --version
Python 3.12.1

$ conda list | grep "cdsetool\|geopandas\|urllib\|pyproj"
cdsetool                  0.2.7                    pypi_0    pypi
geopandas                 0.14.2          py312h06a4308_0
geopandas-base            0.14.2          py312h06a4308_0
pyproj                    3.6.1           py312h6268720_0
urllib3                   2.0.7              pyhd8ed1ab_0    conda-forge
```

**Comment by massimozanetti** [comment]
@pedep I am trying to install `urllib3==2.0.7` in my conda environment, but I don't understand why it is not updating in the list....
```
(gdal) max@asuka:~/sentinel$ conda list | grep "geopandas\|cdsetool\|urllib"
cdsetool                  0.2.3                    pypi_0    pypi
geopandas                 0.12.2           py38h06a4308_0
geopandas-base            0.12.2           py38h06a4308_0
urllib3                   1.26.7                   pypi_0    pypi
(gdal) max@asuka:~/sentinel$ conda install urllib3==2.0.7
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /home/max/anaconda3/envs/gdal

  added / updated specs:
    - urllib3==2.0.7


The following packages will be SUPERSEDED by a higher-priority channel:

  urllib3            conda-forge/noarch::urllib3-2.0.7-pyh~ --> pkgs/main/linux-64::urllib3-2.0.7-py38h06a4308_0


Proceed ([y]/n)? y

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Retrieving notices: ...working... done
(gdal) max@asuka:~/sentinel$ conda list | grep "geopandas\|cdsetool\|urllib"
cdsetool                  0.2.3                    pypi_0    pypi
geopandas                 0.12.2           py38h06a4308_0
geopandas-base            0.12.2           py38h06a4308_0
urllib3                   1.26.7                   pypi_0    pypi
```
I am in the conda env, but after listing the `urllib3` is still the pypi old version... ???

**Comment by pedep** [comment]
@massimozanetti I suspect you have installed urllib3 via pip instead of conda, and pip takes precedence over the conda package. Could you try `pip install urllib3==2.0.7` instead of the `conda install` command?

**Comment by massimozanetti** [comment]
Thank you @pedep. Now @pjonsson here is what I've got. This is my setting now:
```
(gdal) max@asuka:~/sentinel$ conda list | grep "geopandas\|cdsetool\|urllib"
cdsetool                  0.2.3                    pypi_0    pypi
geopandas                 0.12.2           py38h06a4308_0
geopandas-base            0.12.2           py38h06a4308_0
urllib3                   2.0.7            py38h06a4308_0
(gdal) max@asuka:~/sentinel$ python --version
Python 3.8.18
```
And that's whay I get after running the reproducer:
```
Traceback (most recent call last):
  File "reproducer.py", line 21, in <module>
    credentials = Credentials(username, password)
  File "/home/max/anaconda3/envs/gdal/lib/python3.8/site-packages/cdsetool/credentials.py", line 69, in __init__
    self.__ensure_tokens()
  File "/home/max/anaconda3/envs/gdal/lib/python3.8/site-packages/cdsetool/credentials.py", line 135, in __ensure_tokens
    self.__validate_tokens()
  File "/home/max/anaconda3/envs/gdal/lib/python3.8/site-packages/cdsetool/credentials.py", line 145, in __validate_tokens
    key = self.__jwk_client.get_signing_key_from_jwt(self.__access_token)
  File "/home/max/anaconda3/envs/gdal/lib/python3.8/site-packages/jwt/jwks_client.py", line 113, in get_signing_key_from_jwt
    return self.get_signing_key(header.get("kid"))
  File "/home/max/anaconda3/envs/gdal/lib/python3.8/site-packages/jwt/jwks_client.py", line 95, in get_signing_key
    signing_keys = self.get_signing_keys()
  File "/home/max/anaconda3/envs/gdal/lib/python3.8/site-packages/jwt/jwks_client.py", line 82, in get_signing_keys
    jwk_set = self.get_jwk_set(refresh)
  File "/home/max/anaconda3/envs/gdal/lib/python3.8/site-packages/jwt/jwks_client.py", line 79, in get_jwk_set
    return PyJWKSet.from_dict(data)
  File "/home/max/anaconda3/envs/gdal/lib/python3.8/site-packages/jwt/api_jwk.py", line 109, in from_dict
    return PyJWKSet(keys)
  File "/home/max/anaconda3/envs/gdal/lib/python3.8/site-packages/jwt/api_jwk.py", line 102, in __init__
    raise PyJWKSetError(
jwt.exceptions.PyJWKSetError: The JWK Set did not contain any usable keys. Perhaps 'cryptography' is not installed?
```

**Comment by massimozanetti** [comment]
@pedep thanks, `pip` is indeed taking precedence over `conda` (?). I am wondering where conda is installing `urllib3` then.....

By the way, @pjonsson under this setting:
```
$ conda list | grep "geopandas\|cdsetool\|urllib"
cdsetool                  0.2.7                    pypi_0    pypi
geopandas                 0.13.2                   pypi_0    pypi
urllib3                   2.0.7            py38h06a4308_0
```

I've got no errors and all files are downloaded incomplete exactly in the same way as before... so for me updated `urllib3` makes no difference...

**Comment by pjonsson** [comment]
@massimozanetti Ok, so you have a different problem than what I have then. I'm using Python 3.10 on Ubuntu 22.04 LTS. I know there is something with SSL changed between 3.8 and 3.10, but the problem you are seeing could just as well be related to internet connection or something else.

I realize you can't easily change OS, does Conda allow you to (easily) make a separate "environment" with Python 3.10 so you could see if the problem persists there too?

**Comment by massimozanetti** [comment]
@pjonsson installed python=3.10.13 in new env and run the reproducer. This is the error:
```
(py10) max@asuka:~/sentinel$ conda list | grep "geopandas\|cdsetool\|urllib"
cdsetool                  0.2.7                    pypi_0    pypi
geopandas                 0.14.3                   pypi_0    pypi
urllib3                   2.2.1                    pypi_0    pypi
(py10) max@asuka:~/sentinel$ python --version
Python 3.10.13
(py10) max@asuka:~/sentinel$ python reproducer.py
Traceback (most recent call last):
  File "/home/max/anaconda3/envs/py10/lib/python3.10/site-packages/requests/models.py", line 971, in json
    return complexjson.loads(self.text, **kwargs)
  File "/home/max/anaconda3/envs/py10/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/home/max/anaconda3/envs/py10/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/home/max/anaconda3/envs/py10/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/max/sentinel/reproducer.py", line 17, in <module>
    for feature in qfeatures:
  File "/home/max/anaconda3/envs/py10/lib/python3.10/site-packages/cdsetool/query.py", line 28, in __next__
    item = self.feature_query[self.index]
  File "/home/max/anaconda3/envs/py10/lib/python3.10/site-packages/cdsetool/query.py", line 63, in __getitem__
    self.__fetch_features()
  File "/home/max/anaconda3/envs/py10/lib/python3.10/site-packages/cdsetool/query.py", line 69, in __fetch_features
    res = requests.get(self.next_url, timeout=120, proxies=self.proxies).json()
  File "/home/max/anaconda3/envs/py10/lib/python3.10/site-packages/requests/models.py", line 975, in json
    raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

**Comment by pjonsson** [comment]
Oh, I have that one as well! Haven't had time to look at it, but it's a separate issue from this one.

It's already been a long day, so I will look at it tomorrow.

**Comment by pjonsson** [comment]
@massimozanetti many thanks for all your testing on my behalf. Since I just have intermittent failures on my side I have to start things and let them run hoping something will be caught within a week, so it's invaluable to have someone that can reproduce the issue reliably.

**Comment by pjonsson** [comment]
@massimozanetti I think the error you got with 3.10 is some variation of the following:
```
time=2024-03-18T18:55:54 level=ERROR msg="Failed to connect to https://catalogue.dataspace.copernicus.eu: 500 Server Error: Internal Server Error for url: https://catalogue.dataspace.copernicus.eu/resto/api/collections/Sentinel2/search.json?maxRecords=1000&startDate=202....
```
I have made a PR that will check that the response code is good before trying to convert to json, and that will eliminate the cryptic error message.

Since I believe the problem you got with 3.10 is intermittent, can you run the reproducer with 3.10 one more time and see if it works?

**Comment by LattaruoloAndrea** [comment]
Hello, I have been trying to reproduce the same code as @massimozanetti tried, I tested on the following requirements:
- Ubuntu 23.10
- Python3 3.11.6
- cdse 0.2.8
- urllib3 2.2.1
- wifi Eduroam internet (institution: https://www.fbk.eu/en/)
I am using the following script:
```Python
from cdsetool.credentials import Credentials
from cdsetool.query import query_features
from cdsetool.download import download_features, download_feature
from cdsetool.monitor import StatusMonitor

username = ''
password = ''
products_dir = ''

uuid_list = ['f2b18bec-4c8b-5c34-b253-d3e881dabbf8', 'a329664e-afa6-5866-9dc3-6830377906bf', '327905fc-6cf7-55b8-905b-630dc4c66615', '3ea6870a-1061-5db3-854a-a1ea1089ecf9', 'a8691ced-c685-5abb-86c9-f74d822e3b46', 'e3a6273e-d5cb-5f05-88e6-2afbbadb6cfe', '4b36b3c6-1d1c-5c51-91c5-5e3d77de6f3b', '807fd94e-8509-5d55-b1d4-6dfb96beced5', 'f233bb25-d79c-50bf-aca7-b3f7d27b7f83', '5628329c-9899-5814-ac41-aff14fa75443', '15810eaf-f5b5-50bf-89d0-32cf895ca18c', 'bf19b5c3-f0c4-5191-a638-a5ba94160693', '540a7b48-2d78-5ed5-996f-ad81a08ef628', '15a655d3-38c1-59d3-85ba-895afc2819cf', 'f62c9a83-63ba-5e02-9e0b-27cd3f128321', 'af723bbf-513f-5d25-8707-2e093a1a0ad5', 'e29669ab-2d25-5258-8ff6-cdb0c70e3e49', 'ba0f80ab-7acc-5777-8ebd-ce86edcfaad5', 'b86fe089-b65d-5c27-8b7e-c73951009c14', 'e88c4152-4a1f-57f6-bc02-e1cd33eb66de']

#query features
features_list = []
for uuid in uuid_list:
  qfeatures = query_features('Sentinel2',{'identifier': uuid})
  for feature in qfeatures:
    features_list.append(feature)

#istantiate credentials
credentials = Credentials(username, password)
#options
options = {'credentials': credentials,'concurrency': 4,'monitor': StatusMonitor()}

#download products: bulk
downloads = download_features(features_list, products_dir, options)
#downloads are yelded
list(downloads)
```
but I did have different issues, the first one that I found was
![Screenshot from 2024-04-09 09-01-49](https://github.com/SDFIdk/CDSETool/assets/36859220/c9484aba-ac24-462e-ab93-c708906f42d7)
which is due to urllib3 exception.
When I got this error the script didn't allow to download all the file and the process didn't terminate but remained in the following state:
![Screenshot from 2024-04-09 09-07-42](https://github.com/SDFIdk/CDSETool/assets/36859220/51e30c99-bab0-4024-b25d-efaefdbcefbb)
(With the tries that I have made I was never been able to download all 20 the products in one go)

Second problem that I found was before starting of the download I just received this error:
`
Traceback (most recent call last):
  File "/home/mithra/Documents/personal/leet_code/reproducer_max.py", line 16, in <module>
    for feature in qfeatures:
  File "/home/mithra/Documents/personal/leet_code/.env/lib/python3.11/site-packages/cdsetool/query.py", line 28, in __next__
    item = self.feature_query[self.index]
           ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
  File "/home/mithra/Documents/personal/leet_code/.env/lib/python3.11/site-packages/cdsetool/query.py", line 63, in __getitem__
    self.__fetch_features()
  File "/home/mithra/Documents/personal/leet_code/.env/lib/python3.11/site-packages/cdsetool/query.py", line 71, in __fetch_features
    response.raise_for_status()
  File "/home/mithra/Documents/personal/leet_code/.env/lib/python3.11/site-packages/requests/models.py", line 1021, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 503 Server Error: Service Unavailable for url: https://catalogue.dataspace.copernicus.eu/resto/api/collections/Sentinel2/search.json?maxRecords=2000&identifier=327905
` 
maybe this helps with debugging

**Comment by pjonsson** [comment]
I've had the IncompleteRead issues too (and what I was expecting when I requested to try urllib3 v2, but @massimozanetti apparently has identified some other problem). For the IncompleteRead issue, I have a draft that needs to be reworked, hopefully I can get to that next week.

There isn't much we can do when the other end responds with a 500 code, but the 503 issue will be improved by retrying more (#120 and #121).

**Comment by pjonsson** [comment]
@LattaruoloAndrea I just reworked the draft and was lucky enough to immediately run into the problem, so my issue (which I believe is the same as yours) is mitigated by #124. Can you try again with that PR?

**Comment by LattaruoloAndrea** [comment]
@pjonsson  I tried your [#124](https://github.com/SDFIdk/CDSETool/pull/124), and it worked just fine, my only tought was to try and test also the current version 
`Name: cdsetool
Version: 0.2.8
Summary: Tools & CLI for interacting with CDSE product APIs
Home-page: 
Author: 
Author-email: Jacob Vejby <javej@sdfi.dk>
License:
Requires: geopandas, pyjwt, requests, rich, typer
Required-by: `
And today I tried two tests both on the current version 0.2.8 and the #124 but in both cases I was able to download all the data without any errors, both times on the two versions of the library. For sure I cannot tell if the #124 solve the problem or not as far as I understood.

**Comment by massimozanetti** [comment]
@pjonsson @LattaruoloAndrea good news! I just re-installed cdsetool=0.2.8 and files are downloaded all ok... maybe the issue is solved? @pjonsson are you still having issues?

**Comment by pjonsson** [comment]
@massimozanetti I'm glad that things have improved for you, but unfortunately I'm not aware of any changes in 0.2.8 that should make any difference for this issue.

Even if the issue appears to be fixed for now, I suggest we keep this ticket open until @pedep re-appears and merges the current PR queue because that touches code paths that this issue involves. After those changes are merged, it would be great if @massimozanetti tests again so we can be sure that things still work for him.

**Comment by pjonsson** [comment]
@massimozanetti Sorry, I lost track of this. There were a bunch of fixes after 0.2.8 that could affect this issue, does 0.2.11 also seem to work as intended for you?

**Comment by massimozanetti** [comment]
> @massimozanetti Sorry, I lost track of this. There were a bunch of fixes after 0.2.8 that could affect this issue, does 0.2.11 also seem to work as intended for you?

@pjonsson I've just tried out the `0.2.11` and the download works as intended for me, on Ubuntu Server. All products are ok.

I believe it still remains another issue about downloads https://github.com/CDSETool/CDSETool/issues/150 (on windows) opened by @LattaruoloAndrea

**Comment by pjonsson** [comment]
@massimozanetti I fixed an issue where the last few kilobytes of the zip file was missing, so you probably want the latest release (0.2.12) if you still have issues with broken zip files.

**Comment by pjonsson** [comment]
@massimozanetti The windows issue is fixed, can this be closed as completed now?

**Comment by massimozanetti** [comment]
Yes I now experience complete downloads.
